### ISSUE TITLE
Multi Value Measurements

### ISSUE THREAD

**Issue by derpeter** [wger-project/wger#1723]
## Use case

Some measurements e.g. Blood Pressure consist of multiple values. While this could be entered into wger via 2 custom measurement this feels wrong. Especially as the  the date picker in web / app only allows to set a date not a time (not sure if i should open a second request on that)  so the measurements cant be linked together by timestamp. E.g. https://codeberg.org/toz/MediLog medilog is a nice example for these multi value measurements tracked via an app.

## Proposal

Extend the custom measurements to allow to define multiple values per measurement.
Optional: Add some of the common ones as selectable defaults. e.g. https://github.com/wger-project/wger/issues/875 lists already most common ones.

**Comment by Bralc003** [comment]
Hi, I would like to work on this

**Comment by rolandgeider** [comment]
Sure @Bralc003! Note that this needs first a round of brainstorming to think about how this could/should work, how to model this in the db, the UI etc. Feel free to ping me either here or on discord to discuss your ideas

**Comment by Bralc003** [comment]
Thanks @rolandgeider. In terms of modeling this in db I was thinking two options

1. Adding a new table for measurement values that would have multiple entries and link back to the original measurement ID in the measurement table. You would have one measurement ID that links to multiple measurement values for example, if you had blood pressure it would link to values for systolic and diastolic pressure. I think this solution would be good but it would likely require a lot changes across existing places that use measurements. Queries that use measurements would also become more complex since you would need to join the tables to retrieve the data.

2. Another option would be to store the values as JSONB for example blood pressure would be stored as {"systolic": 100, "diastolic": 60}. This would be simpler to implement and to query the measurements from the DB but would be less consistent and clean in terms of how the data is structured.

Let me know what you think of these ideas.

**Comment by rolandgeider** [comment]
Hey! Good question. I think I would tend to go towards the second option, as it's simpler.

However, this would be "its own thing" and we would need two ways to read the value. Also, this would not be compatible to dynamic measurements, which is something that's also being worked on https://github.com/wger-project/wger/issues/1560 (don't really know how that's gonna look like)

We have a similar situation in that exercises that are similar are collected in one group. This was previously done via a FK to separate empty table, but we recently moved this to a [simple field](https://github.com/wger-project/wger/blob/master/wger/exercises/models/base.py#L108), which among other things makes synchronizing and editing or deleting the group much easier. We could do something similar here and give the categories a uuid, and let the frontend combine them?

**Comment by Bralc003** [comment]
I agree, I think this would be a good way to group values together and should be easy to combine them at the front end. 

I've noticed the front end currently displays the category name for the measurement as the title. I think it would be useful to have a group name for multi value measurements. For example you would have a group name stored for "Blood Pressure" which would be displayed with the systolic and diastolic categories.

A good way to do this would be to have a new table that stores the group name and the category table has a FK to the group name.

Otherwise to keep things simpler for now, we could just display all the category names in the title.

**Comment by rolandgeider** [comment]
You are right, hadn't thought about the name. Yes, let's add a new table.

**Comment by Bralc003** [comment]
I think that should be all we need for the DB to handle the multi value measurements.

With the UI the add measurement form would need to be changed to include the group name and additional categories. I was thinking to have by default the form ask the user for a group name and 1 category, then have a button for the user to add more categories.

The line graph would likely display the values for all categories so I was thinking we should limit the amount of categories a user can have for a measurement to 5. Even with limiting the categories to 5 the line graph could potentially feel cluttered when there are a lot of values added.
